### PR TITLE
Dynamic bp tree

### PR DESCRIPTION
**Pull request by kwcantrell**
This PR implements a shear method for the bp tree that is based on https://github.com/wasade/improved-octo-waddle/blob/master/bp/_bp.pyx#L732

@wasade if you can take a look at this that would be awesome!

### CODE REVIEW

File: tests/test-bp-tree.js [hunk 1]
```diff
@@ -877,5 +877,101 @@ require(["jquery", "ByteArray", "BPTree"], function ($, ByteArray, BPTree) {
...
+                "r",
+            ]);
+            deepEqual(result.lengths_, [null, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]);
+
+            keep = new Set(["7", "10", "11"]);
+            result = preShearBPTree.shear(keep);
+            deepEqual(result.b_, [1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0]);
+            deepEqual(result.names_, [null, "7", "10", "11", "9", "8", "r"]);
+            deepEqual(result.lengths_, [null, 6, 7, 8, 9, 10, 11]);
+        });
```
Comment (ElDeveloper) [inline tests/test-bp-tree.js:975]: @kwcantrell can you add a test case for when all tips are removed?

File: empress/support_files/js/bp-tree.js [hunk 2]
```diff
@@ -971,5 +971,64 @@ define(["ByteArray", "underscore"], function (ByteArray, _) {
...
+
+        var newBitArray = [];
+        for (i = 0; i < mask.length; i++) {
+            if (mask[i] !== undefined) {
+                newBitArray.push(mask[i]);
+            }
+
+            // get name and length of node
+            // Note: names and lengths of nodes are stored in postorder
+
```
Comment (ElDeveloper) [inline empress/support_files/js/bp-tree.js:1024]:
```suggestion
```

File: empress/support_files/js/bp-tree.js [hunk 3]
```diff
@@ -971,5 +971,64 @@ define(["ByteArray", "underscore"], function (ByteArray, _) {
         return this._nameToNodes[name];
     };
 
+    /**
+     * tips - Set
```
Comment (ElDeveloper) [inline empress/support_files/js/bp-tree.js:975]: Can you fill the rest of the documentation?

**Review by ElDeveloper — CHANGES_REQUESTED** [review]
Looks great @kwcantrell, I just added a few small comments.

**Comment by wasade** [comment]
This looks awesome!!! +1

File: empress/support_files/js/bp-tree.js (hunk 3 above)
Comment (kwcantrell) [reply empress/support_files/js/bp-tree.js]: doc added

File: tests/test-bp-tree.js (hunk 1 above)
Comment (kwcantrell) [reply tests/test-bp-tree.js:981]: test added

File: empress/support_files/js/bp-tree.js [hunk 4]
```diff
@@ -971,5 +971,71 @@ define(["ByteArray", "underscore"], function (ByteArray, _) {
         return this._nameToNodes[name];
     };
 
+    /**
+     * Returns a new BPTree object that contains just the tips (and ancestors)
+     * of the nodes in keepTips.
+     *
+     * This method was ported from iow.
+     * https://github.com/wasade/improved-octo-waddle/blob/master/bp/_bp.pyx#L732
```
Comment (ElDeveloper) [inline empress/support_files/js/bp-tree.js]: Assuming this was based on the current tip of the main repo:

```suggestion
     * https://github.com/wasade/improved-octo-waddle/blob/master/bp/_bp.pyx#L732
     * SHA: ac16703d2e00eb0abb560a647bff32be120311e1
```

**Review by ElDeveloper — APPROVED** [review]

**Comment by ElDeveloper** [comment]
Thanks @kwcantrell

**Pull request merged**